File: vibeshield-python/examples/deserialization_usage.py
"""
VibeShield Deserialization Protection — Usage Example

Demonstrates how to safely parse JSON payloads, enforce depth checks,
sanitize dangerous keys, and statically analyze code for unsafe deserialization.

Run with: python examples/deserialization_usage.py
"""

import os
import sys

# Allow importing from src/
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from src import (
    safe_json_parse,
    detect_unsafe_deserialization,
    enforce_safe_json,
    VibeShieldDeserializationError
)

print("=============================================================")
print("[SECURE] VIBESHIELD DESERIALIZATION PROTECTION DEMO")
print("=============================================================\n")

# 1. Safe JSON parsing with depth validation and sanitisation
print("--- 1. Using safe_json_parse() ---")
payloads = [
    ('{"a": 1, "b": 2}', 5),
    ('{"a": {"b": {"c": {"d": 4}}}}', 3), # Exceeds depth 3
    ('{"__proto__": {"polluted": true}, "safe_key": "hello"}', 5), # Prototype pollution
    ('{"nested": {"constructor": "danger", "prototype": "danger"}}', 5), # Other keyword sanitisation
    ('{"malformed":', 5) # Malformed JSON
]

for payload, depth in payloads:
    print(f"Parsing: {payload} (Max Depth: {depth})")
    try:
        parsed = safe_json_parse(payload, max_depth=depth)
        print(f"  [SUCCESS] Result: {parsed}")
    except VibeShieldDeserializationError as e:
        print(f"  [BLOCKED] Error: {e}")
print()

# 2. Middleware safe parser
print("--- 2. Using enforce_safe_json() Middleware ---")
body_texts = [
    "",
    "   ",
    '{"username": "vibe_user"}',
    '{"nested": {"nested": {"nested": 1}}}'
]

for body in body_texts:
    print(f"Enforcing safe body: {repr(body)}")
    try:
        parsed = enforce_safe_json(body, max_depth=2)
        print(f"  [SUCCESS] Result: {parsed}")
    except VibeShieldDeserializationError as e:
        print(f"  [BLOCKED] Error: {e}")
print()

# 3. Static Analysis Detector
print("--- 3. Using detect_unsafe_deserialization() Detector ---")
sample_codes = [
    # Safe code
    """
import json
import yaml

def parse(data):
    obj = json.loads(data)
    cfg = yaml.safe_load(data)
    return obj, cfg
""",
    # Unsafe code
    """
import pickle
import yaml
import marshal
import shelve
import dill
import jsonpickle

def do_unsafe(data):
    # Unsafe pickle
    p = pickle.loads(data)
    # Unsafe yaml
    y = yaml.load(data)
    # Unsafe marshal
    m = marshal.loads(data)
    # Unsafe shelve
    s = shelve.open("db.db")
    # Unsafe dill
    d = dill.loads(data)
    # Unsafe jsonpickle
    jp = jsonpickle.decode(data)
""",
    # Safe YAML Loader usage
    """
import yaml
def load_yaml(data):
    return yaml.load(data, Loader=yaml.SafeLoader)
"""
]

for i, code in enumerate(sample_codes, 1):
    print(f"Analyzing Code Block {i}:")
    res = detect_unsafe_deserialization(code)
    if res["safe"]:
        print("  [SAFE] No unsafe deserialization patterns detected.")
    else:
        print(f"  [UNSAFE] Detected {len(res['findings'])} findings:")
        for finding in res["findings"]:
            print(f"    - Line {finding['line']}: pattern '{finding['pattern']}' ({finding['severity']})")
            print(f"      Code: {finding['original_code']}")
            print(f"      Suggestion: {finding['suggestion']}")
print()

File: vibeshield-ts/examples/deserialization-usage.ts
/**
 * VibeShield Deserialization & Prototype Pollution Protection — Usage Example
 *
 * This file demonstrates how to use the Deserialization Protection utilities to:
 * 1. Safe JSON parsing with prototype pollution protection.
 * 2. Deep JSON nesting DoS prevention.
 * 3. Static scanning for dangerous libraries (e.g. node-serialize) and eval execution.
 *
 * Run with: npx tsx examples/deserialization-usage.ts
 */

import {
  safeJsonParse,
  detectUnsafeDeserialization,
  enforceSafeJson,
  VibeShieldDeserializationError
} from '../src/index.js';

console.log('=============================================================');
console.log('🛡️ VIBESHIELD DESERIALIZATION & PROTOTYPE POLLUTION PROTECTION');
console.log('=============================================================\n');

// 1. Prototype Pollution Prevention
console.log('--- 1. Prototype Pollution Prevention ---');
const maliciousPayload = `{
  "user": "Alice",
  "role": "guest",
  "__proto__": {
    "isAdmin": true
  },
  "constructor": {
    "prototype": {
      "isSuperAdmin": true
    }
  }
}`;

console.log('Malicious payload containing prototype pollution:');
console.log(maliciousPayload);

try {
  // Using standard JSON.parse can cause issues depending on how property access is handled
  const normalParsed = JSON.parse(maliciousPayload);
  console.log(`\nStandard JSON.parse:`);
  console.log(`  - user: ${normalParsed.user}`);
  console.log(`  - role: ${normalParsed.role}`);
  console.log(`  - __proto__.isAdmin: ${normalParsed.__proto__?.isAdmin}`);

  // Using safeJsonParse
  const safeParsed = safeJsonParse<any>(maliciousPayload);
  console.log(`\n🛡️ VibeShield safeJsonParse:`);
  console.log(`  - user: ${safeParsed.user}`);
  console.log(`  - role: ${safeParsed.role}`);
  console.log(`  - __proto__: ${safeParsed.__proto__}`); // Should be undefined
  console.log(`  - constructor: ${safeParsed.constructor}`); // Should be undefined
  console.log(`  - Global Object Pollution Check (isAdmin): ${(Object.prototype as any).isAdmin}`); // Should be undefined
} catch (error) {
  console.log(`❌ Parse failed: ${(error as any).message}`);
}
console.log('\n');

// 2. Nesting Depth DoS Protection
console.log('--- 2. Nesting Depth DoS Protection ---');
// Let's create a nested structure
let nestedPayload = '{"value": "leaf"}';
for (let i = 0; i < 70; i++) {
  nestedPayload = `{"level_${i}": ${nestedPayload}}`;
}

console.log('Attempting to parse payload with depth of 71 layers (default max is 64):');
try {
  safeJsonParse(nestedPayload);
  console.log('✅ Parsed successfully? (Should not happen with default 64 max depth!)');
} catch (error) {
  if (error instanceof VibeShieldDeserializationError) {
    console.log(`🛡️ Blocked: ${error.message}`);
  } else {
    console.log(`💥 Unknown Error: ${(error as any).message}`);
  }
}

console.log('\nParsing with custom max depth set to 80:');
try {
  const parsed = safeJsonParse<any>(nestedPayload, { maxDepth: 80 });
  console.log('✅ Successfully parsed deep JSON payload by overriding maxDepth limit!');
} catch (error) {
  console.log(`❌ Failed: ${(error as any).message}`);
}
console.log('\n');

// 3. Middleware Helper enforceSafeJson
console.log('--- 3. Middleware Helper (enforceSafeJson) ---');
const emptyBody = '   ';
console.log(`Parsing empty body: "${emptyBody}"`);
const parsedEmpty = enforceSafeJson(emptyBody);
console.log('Result:', parsedEmpty); // Should be {}

const validBody = '{"active": true}';
console.log(`Parsing valid body: "${validBody}"`);
const parsedValid = enforceSafeJson(validBody);
console.log('Result:', parsedValid);
console.log('\n');

// 4. Static Scanning for Unsafe Deserialization
console.log('--- 4. Static Scanning for Unsafe Deserialization ---');
const vulnerableSourceCode = `
import serialize from 'serialize-to-js';
const express = require('express');
const app = express();

app.post('/unserialize', (req, res) => {
  const payload = req.body.data;
  // Dangerous: eval with unserialize keyword
  const obj = eval("unserialize(" + payload + ")");
  
  // Dangerous: using serialize-to-js directly
  const data = serialize.deserialize(payload);
  
  res.send("Done");
});
`;

console.log('Scanning source code for unsafe deserialization practices...');
const scanResult = detectUnsafeDeserialization(vulnerableSourceCode);
console.log(`Safe status: ${scanResult.safe ? '✅ SAFE' : '❌ VULNERABLE'}`);
console.log(`Findings count: ${scanResult.findings.length}`);
for (const finding of scanResult.findings) {
  console.log(`  - Line ${finding.line}: [${finding.severity.toUpperCase()}] Pattern: "${finding.pattern}"`);
  console.log(`    Code: \`${finding.originalCode}\``);
  console.log(`    Suggestion: ${finding.suggestion}`);
}
console.log('\n=============================================================');
